content/gocourse-2019-05-melbourne.slide

```diff
@@ -813,50 +816,103 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
+- Creation of new error type can be done in a few ways:
 
-The `error` type is a built-in interface:
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
+	}
+- Or use formatted error message using `fmt.Errorf`
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
+	}
+
+* Errors in Go (...cotd)
 
-	type error interface {
-	    Error() string
+- Custom error types can be created. Example from `net` package -
+
+	type Error interface {
+		error
+		Timeout() bool
+		Temporary() bool
 	}
+- Implementation of the `Error` interface here - [[https://golang.org/pkg/net/#AddrError][net/AddrError]]
+- The error values can be logged and printed. Example -
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
-		}
-		return catStore.cats[id], nil
+
+* Error Handling
+
+- Error handling: Check for the presence of an error (compare with nil).
+- Do not compare error variables. ([[https://goplay.space/#mgw9QAmeLUP][Example]])
+- Errors are values so it can be programmed. ([[https://goplay.space/#MTqqxwEdq-A][Example]])
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
 	}
+- Handle errors gracefully ([[https://goplay.space/#kpFwYYBE0ND][A useful pattern]])
+- To check if an `error` is of certain type use custom error types with type assertion
+- Example: ([[https://goplay.space/#t4s_BrAQTLt][Errors and type assertion]])
+- By convention many standard library functions return `error` as the second return value.
 
+"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+
+* Defer
+
+- `defer` (keyword) allows you to postpone execution.
+- When `defer` statement executes, "the parameters to call are evaluated and saved anew"
+- Deferred functions are called when "function returns" or "stack unwinds".
+- Deferred functions are called in reverse order.
+- Used for resource cleanup (close IO resources, release locks etc.).
+- Syntax
+	defer <expression>
+- Example [[https://goplay.space/#wvwUzU1TpNw][How defer works]]
 
-* Errors (2/3)
 
-"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
+* Panic and Recover
 
-Panic should be used when something went unexpectedly wrong.
-For instance, initialise a regular expression with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if the pattern is a string constant.
-`MustCompile` will `panic` at initialisation.
+`panic` and `recover` behaves similar to `try/catch` in other languages.
 
-	var re = regexp.MustCompile(`^[a-c]+$`)
-	var re2 = regexp.MustCompile(`^[a-c+$`) // panic
+- `panic` causes
+- |__ stack unwinding
```

Comment:
This presentation of a nested list is somewhat odd - can you just indent it another level like normal nested lists? Better not to use ASCII drawings for presentation - we moved past that quite some years ago :)

Reply:
Does the ASCII-art mean something in the .slide format?

Reply:
I don't know a way to indent two levels in present. So I resorted to ASCII art :) I've removed it and put them all in bullets at the same level.